In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv("../data/raw/Banking_churn_prediction.csv")
df.head()

,customer_id,vintage,age,gender,dependents,occupation,city,customer_nw_category,branch_code,current_balance,...,average_monthly_balance_prevQ,average_monthly_balance_prevQ2,current_month_credit,previous_month_credit,current_month_debit,previous_month_debit,current_month_balance,previous_month_balance,churn,last_transaction
0,1,2101,66,Male,0.0,self_employed,187.0,2,755,1458.71,...,1458.71,1449.07,0.20,0.20,0.20,0.20,1458.71,1458.71,0,2019-05-21
1,2,2348,35,Male,0.0,self_employed,NaN,2,3214,5390.37,...,7799.26,12419.41,0.56,0.56,5486.27,100.56,6496.78,8787.61,0,2019-11-01
2,4,2194,31,Male,0.0,salaried,146.0,2,41,3913.16,...,4910.17,2815.94,0.61,0.61,6046.73,259.23,5006.28,5070.14,0,NaT
3,5,2329,90,NaN,NaN,self_employed,1020.0,2,582,2291.91,...,2084.54,1006.54,0.47,0.47,0.47,2143.33,2291.91,1669.79,1,2019-08-06
4,6,1579,42,Male,2.0,self_employed,1494.0,3,388,927.72,...,1643.31,1871.12,0.33,714.61,588.62,1538.06,1157.15,1677.16,1,2019-11-03


In [3]:
df.shape

(28382, 21)

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 28382 entries, 0 to 28381
Data columns (total 21 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   customer_id                     28382 non-null  int64  
 1   vintage                         28382 non-null  int64  
 2   age                             28382 non-null  int64  
 3   gender                          27857 non-null  object 
 4   dependents                      25919 non-null  float64
 5   occupation                      28302 non-null  object 
 6   city                            27579 non-null  float64
 7   customer_nw_category            28382 non-null  int64  
 8   branch_code                     28382 non-null  int64  
 9   current_balance                 28382 non-null  float64
 10  previous_month_end_balance      28382 non-null  float64
 11  average_monthly_balance_prevQ   28382 non-null  float64
 12  average_monthly_balance_prevQ2  

In [5]:
df.describe()

,customer_id,vintage,age,dependents,city,customer_nw_category,branch_code,current_balance,previous_month_end_balance,average_monthly_balance_prevQ,average_monthly_balance_prevQ2,current_month_credit,previous_month_credit,current_month_debit,previous_month_debit,current_month_balance,previous_month_balance,churn
count,28382.000000,28382.000000,28382.000000,25919.000000,27579.000000,28382.000000,28382.000000,2.838200e+04,2.838200e+04,2.838200e+04,2.838200e+04,2.838200e+04,2.838200e+04,2.838200e+04,2.838200e+04,2.838200e+04,2.838200e+04,28382.000000
mean,15143.508667,2091.144105,48.208336,0.347236,796.109576,2.225530,925.975019,7.380552e+03,7.495771e+03,7.496780e+03,7.124209e+03,3.433252e+03,3.261694e+03,3.658745e+03,3.339761e+03,7.451133e+03,7.495177e+03,0.185329
std,8746.454456,272.676775,17.807163,0.997661,432.872102,0.660443,937.799129,4.259871e+04,4.252935e+04,4.172622e+04,4.457581e+04,7.707145e+04,2.968889e+04,5.198542e+04,2.430111e+04,4.203394e+04,4.243198e+04,0.388571
min,1.000000,73.000000,1.000000,0.000000,0.000000,1.000000,1.000000,-5.503960e+03,-3.149570e+03,1.428690e+03,-1.650610e+04,1.000000e-02,1.000000e-02,1.000000e-02,1.000000e-02,-3.374180e+03,-5.171920e+03,0.000000
25%,7557.250000,1958.000000,36.000000,0.000000,409.000000,2.000000,176.000000,1.784470e+03,1.906000e+03,2.180945e+03,1.832507e+03,3.100000e-01,3.300000e-01,4.100000e-01,4.100000e-01,1.996765e+03,2.074407e+03,0.000000
50%,15150.500000,2154.000000,46.000000,0.000000,834.000000,2.000000,572.000000,3.281255e+03,3.379915e+03,3.542865e+03,3.359600e+03,6.100000e-01,6.300000e-01,9.193000e+01,1.099600e+02,3.447995e+03,3.465235e+03,0.000000
75%,22706.750000,2292.000000,60.000000,0.000000,1096.000000,3.000000,1440.000000,6.635820e+03,6.656535e+03,6.666887e+03,6.517960e+03,7.072725e+02,7.492350e+02,1.360435e+03,1.357553e+03,6.667958e+03,6.654693e+03,0.000000
max,30301.000000,2476.000000,90.000000,52.000000,1649.000000,3.000000,4782.000000,5.905904e+06,5.740439e+06,5.700290e+06,5.010170e+06,1.226985e+07,2.361808e+06,7.637857e+06,1.414168e+06,5.778185e+06,5.720144e+06,1.000000


In [6]:
df.isnull().sum()

customer_id                          0
vintage                              0
age                                  0
gender                             525
dependents                        2463
occupation                          80
city                               803
customer_nw_category                 0
branch_code                          0
current_balance                      0
previous_month_end_balance           0
average_monthly_balance_prevQ        0
average_monthly_balance_prevQ2       0
current_month_credit                 0
previous_month_credit                0
current_month_debit                  0
previous_month_debit                 0
current_month_balance                0
previous_month_balance               0
churn                                0
last_transaction                     0
dtype: int64

In [7]:
duplicates = df.duplicated().sum()
print("duplicate rows:", duplicates)

df = df.drop_duplicates()

duplicate rows: 0


## Step 1: Duplicate Check
No duplicate rows found / Removed X duplicate rows.

In [8]:
df["last_transaction"] = pd.to_datetime(df["last_transaction"], errors="coerce")

In [9]:
df["last_transaction"].isnull().sum()

3223

## step 2: last_transaction Cleaning Log

Converted column to datetime format.
3223 rows contained missing/invalid dates and were converted to NaT.
Values preserved for future churn behavior analysis.

In [10]:
round(df["last_transaction"].isnull().sum() / len(df) * 100, 2)

11.36

In [11]:
df.isnull().sum()

customer_id                          0
vintage                              0
age                                  0
gender                             525
dependents                        2463
occupation                          80
city                               803
customer_nw_category                 0
branch_code                          0
current_balance                      0
previous_month_end_balance           0
average_monthly_balance_prevQ        0
average_monthly_balance_prevQ2       0
current_month_credit                 0
previous_month_credit                0
current_month_debit                  0
previous_month_debit                 0
current_month_balance                0
previous_month_balance               0
churn                                0
last_transaction                  3223
dtype: int64

In [12]:
df["gender"] = df["gender"].fillna(df["gender"].mode()[0])
df["occupation"] = df["occupation"].fillna(df["occupation"].mode()[0])
df["dependents"] = df["dependents"].fillna(df["dependents"].median())
df["city"] = df["city"].fillna(df["city"].mode()[0])

## Step 3: Missing Values Treated
- gender → mode
- occupation → mode
- dependents → median
- city → mode

In [13]:
df.isnull().sum()

customer_id                          0
vintage                              0
age                                  0
gender                               0
dependents                           0
occupation                           0
city                                 0
customer_nw_category                 0
branch_code                          0
current_balance                      0
previous_month_end_balance           0
average_monthly_balance_prevQ        0
average_monthly_balance_prevQ2       0
current_month_credit                 0
previous_month_credit                0
current_month_debit                  0
previous_month_debit                 0
current_month_balance                0
previous_month_balance               0
churn                                0
last_transaction                  3223
dtype: int64

In [14]:
df.duplicated().sum()

0

In [17]:
df["age"].describe()
df["dependents"].describe()

count    28382.000000
mean         0.317102
std          0.958386
min          0.000000
25%          0.000000
50%          0.000000
75%          0.000000
max         52.000000
Name: dependents, dtype: float64

In [18]:
df["gender"] = df["gender"].fillna(df["gender"].mode()[0])
df["occupation"] = df["occupation"].fillna(df["occupation"].mode()[0])
df["dependents"] = df["dependents"].fillna(df["dependents"].median())
df["city"] = df["city"].fillna(df["city"].mode()[0])

In [19]:
df["dependents"] = df["dependents"].astype(int)
df["city"] = df["city"].astype(int)

In [20]:
df.isnull().sum()

customer_id                          0
vintage                              0
age                                  0
gender                               0
dependents                           0
occupation                           0
city                                 0
customer_nw_category                 0
branch_code                          0
current_balance                      0
previous_month_end_balance           0
average_monthly_balance_prevQ        0
average_monthly_balance_prevQ2       0
current_month_credit                 0
previous_month_credit                0
current_month_debit                  0
previous_month_debit                 0
current_month_balance                0
previous_month_balance               0
churn                                0
last_transaction                  3223
dtype: int64

In [22]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 28382 entries, 0 to 28381
Data columns (total 21 columns):
 #   Column                          Non-Null Count  Dtype         
---  ------                          --------------  -----         
 0   customer_id                     28382 non-null  int64         
 1   vintage                         28382 non-null  int64         
 2   age                             28382 non-null  int64         
 3   gender                          28382 non-null  object        
 4   dependents                      28382 non-null  int64         
 5   occupation                      28382 non-null  object        
 6   city                            28382 non-null  int64         
 7   customer_nw_category            28382 non-null  int64         
 8   branch_code                     28382 non-null  int64         
 9   current_balance                 28382 non-null  float64       
 10  previous_month_end_balance      28382 non-null  float64       
 11  av

In [23]:
df["age"].describe()

count    28382.000000
mean        48.208336
std         17.807163
min          1.000000
25%         36.000000
50%         46.000000
75%         60.000000
max         90.000000
Name: age, dtype: float64

In [24]:
df["dependents"].describe()

count    28382.000000
mean         0.317102
std          0.958386
min          0.000000
25%          0.000000
50%          0.000000
75%          0.000000
max         52.000000
Name: dependents, dtype: float64

In [25]:
df[df["age"] < 18][["customer_id", "age", "churn"]].head(20)

,customer_id,age,churn
34,37,9,0
44,47,17,0
47,50,4,0
63,67,13,0
115,120,10,1
133,141,3,0
200,216,3,0
259,281,15,0
292,315,6,1
411,441,14,0


In [26]:
(df["age"] < 18).sum()

806

In [27]:
df[df["dependents"] > 10][["customer_id", "dependents", "age", "occupation"]]

,customer_id,dependents,age,occupation
3741,3988,52,43,self_employed
7820,8331,36,41,salaried
9082,9688,50,58,salaried
14016,14957,25,30,self_employed
23681,25267,32,45,salaried


## people with age below 18 can have student account or guardianship account or junior savings account

## dependencies seem odd


In [28]:
df.loc[df["dependents"] > 10, "dependents"] = df['dependents'].median()

In [29]:
df["dependents"] = df["dependents"].astype(int)

## Anomaly Handling

Age values below 18 were retained because junior/minor bank accounts may exist.

5 unrealistic dependents values (>10) were identified as probable data entry errors and replaced using the median dependents value.

In [30]:
df["dependents"].describe()

count    28382.000000
mean         0.310232
std          0.797299
min          0.000000
25%          0.000000
50%          0.000000
75%          0.000000
max          9.000000
Name: dependents, dtype: float64

In [31]:
df.duplicated().sum()

0

In [32]:
df.to_csv("../data/processed/cleaned_banking_churn.csv", index=False)

# Final Data Cleaning Summary

## Objective
Prepare the raw banking churn dataset for analysis by identifying and resolving data quality issues, standardizing formats, and exporting a clean version for downstream EDA, statistical analysis, and dashboarding.

## Dataset Overview
- Original Shape: **28,382 rows × 21 columns**
- Source File: `data/raw/Banking_churn_prediction.csv`

## Cleaning Steps Performed

### 1. Initial Data Audit
Performed:
- Dataset shape review
- Column datatype inspection
- Descriptive statistics
- Missing value assessment
- Duplicate row check

### 2. Duplicate Records
- Checked duplicate rows.
- No duplicate rows found / duplicates removed if detected.

### 3. Datatype Corrections
- Converted `last_transaction` from object to datetime format using safe parsing.

### 4. Missing Value Treatment
Handled null values as follows:

- `gender` → filled using mode
- `occupation` → filled using mode
- `dependents` → filled using median
- `city` → filled using mode

### 5. Date Column Handling
- `last_transaction` conversion produced **3,223 missing values (NaT)** due to blank/invalid dates.
- Missing dates were retained for future transaction-recency analysis.

### 6. Outlier / Anomaly Review

#### Age
- Minimum age observed: 1
- Ages below 18 were retained, as junior/minor banking accounts may exist.

#### Dependents
- Unrealistic values (>10) detected.
- Extreme values were replaced using median dependents value.

### 7. Format Standardization
- Converted `dependents` to integer datatype.
- Converted `city` to integer datatype.

## Final Null Value Status
- All columns cleaned successfully.
- Only `last_transaction` retains missing values intentionally.

## Final Output
Cleaned dataset exported to:

```text
data/processed/cleaned_banking_churn.csv